### PR TITLE
Export more rocksdb statistics to the info command

### LINKED ISSUE #1049: Export the block cache hit/miss statistics
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Motivation

Block cache hit/miss is the key metric for the read performance, so it can help a lot when profiling if we can export the block cache hit/miss statistics to users.

### Solution

Read the block cache statistics from RocksDB, then export through the info command.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by xiaobiaozhao**
close https://github.com/apache/incubator-kvrocks/issues/1049

### CODE REVIEW

File: src/stats/stats.h [hunk 1]
```diff
@@ -37,6 +37,51 @@ enum StatsMetricFlags {
...
+  STATS_METRIC_BLOOM_FILTER_PREFIX_USEFUL,
+
+  STATS_METRIC_COMPACT_READ_BYTES,   // Bytes read during compaction
+  STATS_METRIC_COMPACT_WRITE_BYTES,  // Bytes written during compaction
+  STATS_METRIC_FLUSH_WRITE_BYTES,    // Bytes written during flush
+
+  // Row cache.
+  STATS_METRIC_ROW_CACHE_HIT,
+  STATS_METRIC_ROW_CACHE_MISS,
+
```
Comment (git-hulk) [inline src/stats/stats.h]: Remove them, we didn't use row cache row.

File: src/stats/stats.h [hunk 2]
```diff
@@ -37,6 +37,51 @@ enum StatsMetricFlags {
...
+  STATS_METRIC_GET_HIT_L2_AND_UP,
+  STATS_METRIC_COMPACTION_KEY_DROP_NEWER_ENTRY,         // key was written with a newer value.
+                                                        // Also includes keys dropped for range del.
+  STATS_METRIC_COMPACTION_KEY_DROP_OBSOLETE,            // The key is obsolete.
+  STATS_METRIC_COMPACTION_KEY_DROP_RANGE_DEL,           // key was covered by a range tombstone.
+  STATS_METRIC_COMPACTION_KEY_DROP_USER,                // user compaction function has dropped the key.
+  STATS_METRIC_COMPACTION_RANGE_DEL_DROP_OBSOLETE,      // all keys in range were deleted.
+  STATS_METRIC_COMPACTION_OPTIMIZED_DEL_DROP_OBSOLETE,  // Deletions obsoleted before bottom level due to file gap
+                                                        // optimization.
+  STATS_METRIC_COMPACTION_CANCELLED,                    // If a compaction was canceled in sfm to prevent ENOSPC
```
Comment (git-hulk) [inline src/stats/stats.h:61]: Why do we need to care about this stat?

File: src/stats/stats.h [hunk 3]
```diff
@@ -37,6 +37,51 @@ enum StatsMetricFlags {
...
+  STATS_METRIC_BYTES_WRITTEN,         // The number of uncompressed bytes issued by DB::Put(), DB::Delete(),
+                                      // DB::Merge(), and DB::Write().
+  STATS_METRIC_BYTES_READ,            // The number of uncompressed bytes read from DB::Get().  It could be
+                                      // either from memtables, cache, or table files.
+                                      // For the number of logical bytes read from DB::MultiGet(),
+                                      // please use NUMBER_MULTIGET_BYTES_READ.
+  STATS_METRIC_NUMBER_DB_SEEK_FOUND,  // The number of calls to seek/next/prev that returned data
+  STATS_METRIC_NUMBER_DB_NEXT_FOUND,
+  STATS_METRIC_NUMBER_DB_PREV_FOUND,
```
Comment (git-hulk) [inline src/stats/stats.h:70]: duplicated, please take a look at the old metrics we have.

File: src/stats/stats.h [hunk 4]
```diff
@@ -37,6 +37,51 @@ enum StatsMetricFlags {
...
+  STATS_METRIC_BLOCK_CACHE_MISS,
+  STATS_METRIC_BLOCK_CACHE_HIT,
+  STATS_METRIC_BLOCK_CACHE_ADD,
+  STATS_METRIC_BLOCK_CACHE_BYTES_READ,
+  STATS_METRIC_BLOCK_CACHE_BYTES_WRITE,
+  STATS_METRIC_BLOOM_FILTER_USEFUL,
+  STATS_METRIC_BLOOM_FILTER_FULL_POSITIVE,
+  STATS_METRIC_BLOOM_FILTER_FULL_TRUE_POSITIVE,
+  STATS_METRIC_PERSISTENT_CACHE_HIT,
+  STATS_METRIC_PERSISTENT_CACHE_MISS,
```
Comment (git-hulk) [inline src/stats/stats.h]: We have no persistent cache now, so it makes no sense to add those two.

File: src/stats/stats.h (hunk 3 above)
Comment (xiaobiaozhao) [reply src/stats/stats.h:70]: Do you mean NUMBER_DB_SEEK?

File: src/stats/stats.h (hunk 2 above)
Comment (xiaobiaozhao) [reply src/stats/stats.h:61]: It  planned for TS

File: src/stats/stats.h (hunk 4 above)
Comment (xiaobiaozhao) [reply src/stats/stats.h]: done

File: src/stats/stats.h (hunk 1 above)
Comment (xiaobiaozhao) [reply src/stats/stats.h]: done

**Comment by git-hulk** [comment]
@xiaobiaozhao Looks like you wrongly merged the previous commits and they appeared as new code. Please fix them first.

**Review by tisonkun — COMMENTED** [review]
You should merge or rebase onto unstable instead of pick unstable commit thus they're different commit object to show in this patch.

File: src/stats/stats.h (hunk 3 above)
Comment (git-hulk) [reply src/stats/stats.h:70]: Yes, I think those three metrics should no mean a lot for us.

**Comment by xiaobiaozhao** [comment]
> You should merge or rebase onto unstable instead of pick unstable commit thus they're different commit object to show in this patch.

done

**Comment by xiaobiaozhao** [comment]
> @xiaobiaozhao Looks like you wrongly merged the previous commits and they appeared as new code. Please fix them first.

done

**Comment by xiaobiaozhao** [comment]
@git-hulk @tisonkun Merge?

**Comment by ShooterIT** [comment]
does this change damage performance?

maybe we can add a new section which include these statistics, only show these when calling `info all` or `info $new_section`

**Comment by git-hulk** [comment]
> does this change damage performance?
> 
> maybe we can add a new section which include these statistics, only show these when calling `info all` or `info $new_section`

Those metrics should NOT affect the performance because there're from memory statistics and are already in the RocksDB section.

**Comment by xiaobiaozhao** [comment]
> > does this change damage performance?
> > maybe we can add a new section which include these statistics, only show these when calling `info all` or `info $new_section`
> 
> Those metrics should NOT affect the performance because there're from memory statistics and are already in the RocksDB section.

I think it's time to merge  @git-hulk @ShooterIT

**Comment by ShooterIT** [comment]
I don't think current change is OK, the output of `info rocksdb` is tool long, it is not easy to read,
i prefer to show these info when calling `info all`, or add a new section to include them

**Comment by git-hulk** [comment]
> I don't think current change is OK, the output of `info rocksdb` is tool long, it is not easy to read, i prefer to show these info when calling `info all`, or add a new section to include them

Another choice maybe to add another command to export those statistics, dose it sound good?

**Review by PragmaTwice — COMMENTED** [review]
I think maybe we need a better way to oganize code about metric track & output, otherwise many repeated code will be added in `server.cc` to do this task for new metrics 🤣 , OR we can move them to `stats.cc`.

**Comment by xiaobiaozhao** [comment]
> I don't think current change is OK, the output of `info rocksdb` is tool long, it is not easy to read, i prefer to show these info when calling `info all`, or add a new section to include them

Maybe the new section could be called "rocksdb-detail"

**Comment by ShooterIT** [comment]
i found we already have "stats" command, it can report lots of rocksdb info.

**Comment by xiaobiaozhao** [comment]
> i found we already have "stats" command, it can report lots of rocksdb info.

@git-hulk What do you think?

**Comment by git-hulk** [comment]
Yes, can move them into the stat command

**Comment by ShooterIT** [comment]
Hi @xiaobiaozhao did you see `stats` command output? it reports lots of info, is it enough?

**Comment by xiaobiaozhao** [comment]
> Hi @xiaobiaozhao did you see `stats` command output? it reports lots of info, is it enough?

I'll take a look at that this weekend

**Comment by xiaobiaozhao** [comment]
I think `stats` is enough

**Pull request closed without merging**